// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/layout-editor/layout-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { LayoutEditor } from '../layout-editor.model';
export class LayoutEditorComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.done = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.elementRef.nativeElement.classList.add('card', 'card-row', 'col-8', 'col-sm-6', 'popup');
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        const /** @type {?} */ width = this.elementRef.nativeElement
            .getBoundingClientRect()
            .width;
        this.elementRef.nativeElement.style.top = (this.layoutEditor.yPosition - 20) + 'px';
        this.elementRef.nativeElement.style.left = (this.layoutEditor.xPosition - width + 20) + 'px';
    }
    /**
     * @return {?}
     */
    onDoneClick() {
        this.done.emit();
    }
}
LayoutEditorComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widgets-layout-editor',
                template: `<div class='title'>Layout Editor</div>
<p>
  <b>Coming Soon!</b>
</p>
<p>
  This will allow some users to edit pages and make custom layouts targeted for specific roles in your organization.
</p>
<p>
  For now, changes made will not be saved.
</p>
<div class='button-container'>
  <button class='btn btn-primary' (click)='onDoneClick()'>Done</button>
</div>`,
                styles: [`:host{position:absolute;z-index:100;background:#fff}`]
            },] },
];
/** @nocollapse */
LayoutEditorComponent.ctorParameters = () => [
    { type: ElementRef, },
];
LayoutEditorComponent.propDecorators = {
    "layoutEditor": [{ type: Input },],
    "done": [{ type: Output },],
};
function LayoutEditorComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    LayoutEditorComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    LayoutEditorComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    LayoutEditorComponent.propDecorators;
    /** @type {?} */
    LayoutEditorComponent.prototype.layoutEditor;
    /** @type {?} */
    LayoutEditorComponent.prototype.done;
    /** @type {?} */
    LayoutEditorComponent.prototype.elementRef;
}
