// @navihealth/architect-list-widget/esm5/app/app.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef } from '@angular/core';
import { BusFactory } from '@navihealth/browser-bus';
var AppComponent = /** @class */ (function () {
    function AppComponent(elementRef) {
        this.elementRef = elementRef;
        this.title = 'architects';
        this.architects = [
            'Seshan',
            'Aneesh',
            'Nijil'
        ];
        this.bus = (/** @type {?} */ (BusFactory)).create();
    }
    /**
     * @param {?} event
     * @return {?}
     */
    AppComponent.prototype.newArchitectOnKeyUp = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        this.newArchitectName = event.target.value;
    };
    /**
     * @return {?}
     */
    AppComponent.prototype.addOnClick = /**
     * @return {?}
     */
    function () {
        if (this.newArchitectName) {
            this.architects.push(this.newArchitectName);
            this.notifyArchitectSelected(this.newArchitectName);
        }
    };
    /**
     * @param {?} architect
     * @return {?}
     */
    AppComponent.prototype.architectOnClick = /**
     * @param {?} architect
     * @return {?}
     */
    function (architect) {
        this.notifyArchitectSelected(architect);
    };
    /**
     * @return {?}
     */
    AppComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.classList.add('architect-list-widget');
    };
    /**
     * @param {?} architect
     * @return {?}
     */
    AppComponent.prototype.notifyArchitectSelected = /**
     * @param {?} architect
     * @return {?}
     */
    function (architect) {
        this.bus.notify({
            eventName: 'architect-selected',
            message: {
                name: architect
            }
        });
    };
    AppComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-architect-list-widget',
                    template: "<div class='col-xs-12'>\n  <input (keyup)='newArchitectOnKeyUp($event)' placeholder='enter name' />\n  <button (click)='addOnClick()' class='btn btn-primary'>Add Architect</button>\n  <ul>\n    <li *ngFor='let architect of architects' (click)='architectOnClick(architect)'>{{ architect }}</li>\n  </ul>\n</div>",
                    styles: [":host.architect-list-widget ul{margin-top:10px;padding:0}:host.architect-list-widget li{list-style:none;color:green}"]
                },] },
    ];
    /** @nocollapse */
    AppComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    return AppComponent;
}());
export { AppComponent };
function AppComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    AppComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    AppComponent.ctorParameters;
    /** @type {?} */
    AppComponent.prototype.title;
    /** @type {?} */
    AppComponent.prototype.architects;
    /** @type {?} */
    AppComponent.prototype.newArchitectName;
    /** @type {?} */
    AppComponent.prototype.bus;
    /** @type {?} */
    AppComponent.prototype.elementRef;
}
