// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/library.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ILibraryWidget() { }
function ILibraryWidget_tsickle_Closure_declarations() {
    /** @type {?} */
    ILibraryWidget.prototype.isRendered;
    /** @type {?} */
    ILibraryWidget.prototype.widget;
}
export class Library {
    /**
     * @param {?} renderedWidgets
     * @param {?} availableWidgets
     */
    constructor(renderedWidgets, availableWidgets) {
        this.renderedWidgets = renderedWidgets;
        this.availableWidgets = availableWidgets;
        this.isDisplayed = false;
        this.buildLibrary();
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    isWidgetRendered(widget) {
        return this.renderedWidgets
            .map(renderedWidget => renderedWidget.id)
            .indexOf(widget.id) >= 0;
    }
    /**
     * @return {?}
     */
    buildLibrary() {
        this.widgets = this.availableWidgets.map(widget => {
            const /** @type {?} */ isRendered = this.isWidgetRendered(widget);
            return { isRendered, widget };
        });
    }
    /**
     * @param {?} xPosition
     * @param {?} yPosition
     * @return {?}
     */
    edit(xPosition, yPosition) {
        this.isDisplayed = true;
        this.xPosition = xPosition;
        this.yPosition = yPosition;
    }
    /**
     * @return {?}
     */
    hide() {
        this.isDisplayed = false;
        this.xPosition = null;
        this.yPosition = null;
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    renderWidget(widget) {
        this.renderedWidgets.push(widget);
        this.buildLibrary();
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    derenderWidget(widget) {
        this.renderedWidgets
            .filter(x => x.id === widget.id)
            .forEach(x => {
            this.renderedWidgets.splice(this.renderedWidgets.indexOf(x), 1);
        });
        this.buildLibrary();
    }
}
function Library_tsickle_Closure_declarations() {
    /** @type {?} */
    Library.prototype.isDisplayed;
    /** @type {?} */
    Library.prototype.xPosition;
    /** @type {?} */
    Library.prototype.yPosition;
    /** @type {?} */
    Library.prototype.widgets;
    /** @type {?} */
    Library.prototype.renderedWidgets;
    /** @type {?} */
    Library.prototype.availableWidgets;
}
