// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/widget.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function IWidth() { }
function IWidth_tsickle_Closure_declarations() {
    /** @type {?} */
    IWidth.prototype.xs;
    /** @type {?|undefined} */
    IWidth.prototype.sm;
    /** @type {?|undefined} */
    IWidth.prototype.md;
    /** @type {?|undefined} */
    IWidth.prototype.lg;
    /** @type {?|undefined} */
    IWidth.prototype.xl;
}
export class Widget {
    /**
     * @param {?} id
     * @param {?} title
     * @param {?} currentWidth
     * @param {?} minimumWidth
     * @param {?} initialHeight
     * @param {?} render
     * @param {?} isHidden
     */
    constructor(id, title, currentWidth, minimumWidth, initialHeight, render, isHidden) {
        this.currentWidth = currentWidth;
        this.minimumWidth = minimumWidth;
        this.isExpanded = true;
        this.isFullScreen = false;
        this.id = id;
        this.title = title;
        this.initialHeight = initialHeight;
        this.render = (/** @type {?} */ (render)).bind(this);
        this.isHidden = isHidden;
        this.currentWidth = currentWidth;
        // This uses Object.assign to create a new object from the one passed in.
        // This is to prevent the instantiator from passing in one object to both
        // currentWidth and minimumWidth and then having that object ref assigned
        // to both properties. If that were to happen, then updating the
        // currentWidth would also update the minimumWidth. The minimumWidth
        // should never be changed once set.
        this.minimumWidth = Object.assign({}, minimumWidth);
        Object.freeze(this.minimumWidth);
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    static createAvailableWidget({ id, title, width, initialHeight, render }) {
        return new Widget(id, title, width, width, initialHeight, render, false);
    }
    /**
     * @param {?} __0
     * @return {?}
     */
    static createRenderedWidget({ id, title, currentWidth, minimumWidth, initialHeight, render }) {
        return new Widget(id, title, currentWidth, minimumWidth, initialHeight, render, false);
    }
    /**
     * @param {?} breakpoint
     * @return {?}
     */
    getCurrentWidth(breakpoint) {
        return this.currentWidth[breakpoint];
    }
    /**
     * @return {?}
     */
    getCurrentWidths() {
        return this.currentWidth;
    }
    /**
     * @param {?} breakpoint
     * @return {?}
     */
    getMinimumWidth(breakpoint) {
        return this.minimumWidth[breakpoint];
    }
    /**
     * @param {?} breakpoint
     * @param {?} width
     * @return {?}
     */
    makeFullWidth(breakpoint, width) {
        if (this.currentWidth[breakpoint]) {
            this.currentWidth[breakpoint] = width;
        }
    }
    /**
     * @param {?} breakpoint
     * @param {?} width
     * @return {?}
     */
    resize(breakpoint, width) {
        this.currentWidth[breakpoint] = width;
    }
    /**
     * @return {?}
     */
    expand() {
        this.isExpanded = true;
    }
    /**
     * @return {?}
     */
    collapse() {
        this.isExpanded = false;
    }
    /**
     * @return {?}
     */
    toggleFullScreen() {
        this.isFullScreen = !this.isFullScreen;
    }
    /**
     * @return {?}
     */
    clone() {
        return new Widget(this.id, this.title, this.currentWidth, this.minimumWidth, this.initialHeight, this.render, this.isHidden);
    }
}
function Widget_tsickle_Closure_declarations() {
    /** @type {?} */
    Widget.prototype.id;
    /** @type {?} */
    Widget.prototype.title;
    /** @type {?} */
    Widget.prototype.initialHeight;
    /** @type {?} */
    Widget.prototype.render;
    /** @type {?} */
    Widget.prototype.isHidden;
    /** @type {?} */
    Widget.prototype.isExpanded;
    /** @type {?} */
    Widget.prototype.isFullScreen;
    /** @type {?} */
    Widget.prototype.currentWidth;
    /** @type {?} */
    Widget.prototype.minimumWidth;
}
