// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/drop-target/drop-target.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, ElementRef } from '@angular/core';
import { Widget } from '../widget.model';
import { BootstrapService } from '../bootstrap.service';
export class DropTargetComponent {
    /**
     * @param {?} elementRef
     * @param {?} bootstrapService
     */
    constructor(elementRef, bootstrapService) {
        this.elementRef = elementRef;
        this.bootstrapService = bootstrapService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.elementRef.nativeElement.style.height = this.height + 'px';
        const /** @type {?} */ classes = this.bootstrapService.getClasses(this.dropTarget);
        this.elementRef.nativeElement.classList.add(...classes);
    }
}
DropTargetComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widget-drop-target',
                template: `<div class='card card-row'></div>`,
                styles: [`:host>div{background:#d3d3d3;height:100%;margin:0}`]
            },] },
];
/** @nocollapse */
DropTargetComponent.ctorParameters = () => [
    { type: ElementRef, },
    { type: BootstrapService, },
];
DropTargetComponent.propDecorators = {
    "dropTarget": [{ type: Input },],
    "height": [{ type: Input },],
};
function DropTargetComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DropTargetComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DropTargetComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    DropTargetComponent.propDecorators;
    /** @type {?} */
    DropTargetComponent.prototype.dropTarget;
    /** @type {?} */
    DropTargetComponent.prototype.height;
    /** @type {?} */
    DropTargetComponent.prototype.elementRef;
    /** @type {?} */
    DropTargetComponent.prototype.bootstrapService;
}
