// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/panel-controls/panel-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Output, ElementRef, EventEmitter } from '@angular/core';
export class PanelControlsComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.editLibrary = new EventEmitter();
        this.editLayout = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.elementRef.nativeElement.classList.add('col-12');
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onLibraryClick(event) {
        this.editLibrary.emit(event);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    onLayoutClick(event) {
        this.editLayout.emit(event);
    }
}
PanelControlsComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widgets-panel-controls',
                template: `<span class='no-highlighting'>
  <span class='material-icons' (click)='onLibraryClick($event)'>add</span>
<span class='material-icons' (click)='onLayoutClick($event)'>dashboard</span>
</span>`,
                styles: [`:host{margin-top:10px;margin-right:20px;text-align:right}`]
            },] },
];
/** @nocollapse */
PanelControlsComponent.ctorParameters = () => [
    { type: ElementRef, },
];
PanelControlsComponent.propDecorators = {
    "editLibrary": [{ type: Output },],
    "editLayout": [{ type: Output },],
};
function PanelControlsComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    PanelControlsComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    PanelControlsComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    PanelControlsComponent.propDecorators;
    /** @type {?} */
    PanelControlsComponent.prototype.editLibrary;
    /** @type {?} */
    PanelControlsComponent.prototype.editLayout;
    /** @type {?} */
    PanelControlsComponent.prototype.elementRef;
}
