// @navihealth/widget-dashboard-ng5/esm5/navihealth-widget-dashboard-ng5.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * Generated bundle index. Do not edit.
 */
export { DashboardModule, Widget } from './public_api';
export { DashboardComponent as ɵa } from './src/dashboard/dashboard.component';
export { BootstrapService as ɵc } from './src/dashboard/widgets/bootstrap.service';
export { DropTargetComponent as ɵi } from './src/dashboard/widgets/drop-target/drop-target.component';
export { LayoutEditorComponent as ɵk } from './src/dashboard/widgets/layout-editor/layout-editor.component';
export { LibraryComponent as ɵh } from './src/dashboard/widgets/library/library.component';
export { PanelControlsComponent as ɵe } from './src/dashboard/widgets/panel-controls/panel-controls.component';
export { PanelComponent as ɵb } from './src/dashboard/widgets/panel/panel.component';
export { ResizeGuidelineComponent as ɵj } from './src/dashboard/widgets/resize-guideline/resize-guideline.component';
export { SpacerComponent as ɵd } from './src/dashboard/widgets/spacer/spacer.component';
export { WidgetControlsComponent as ɵg } from './src/dashboard/widgets/widget-controls/widget-controls.component';
export { WidgetComponent as ɵf } from './src/dashboard/widgets/widget/widget.component';
