// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/widget-controls/widget-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { Widget } from '../widget.model';
export class WidgetControlsComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.close = new EventEmitter();
        this.fullscreen = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.elementRef.nativeElement.classList.add('col-12');
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        // We'll set the width of the title based on the width left by the buttons.
        // This will make the overflow on the title work properly. But we won't
        // do this until the stack is empty. Without the setTime, the first page load
        // after a Hard Reload does not calculate the offsetWidths correctly.
        setTimeout(() => {
            const /** @type {?} */ containerElement = this.elementRef.nativeElement.children[0];
            const /** @type {?} */ controlsWidth = containerElement.offsetWidth;
            const /** @type {?} */ buttonsElement = containerElement.querySelector('.buttons');
            const /** @type {?} */ buttonsWidth = buttonsElement.offsetWidth;
            const /** @type {?} */ padding = 8;
            const /** @type {?} */ titleElement = containerElement.querySelector('.title');
            titleElement.style.width = (100 - padding - (buttonsWidth / controlsWidth * 100)) + '%';
        }, 0);
    }
    /**
     * @return {?}
     */
    onCollapseClick() {
        this.widget.collapse();
    }
    /**
     * @return {?}
     */
    onExpandClick() {
        this.widget.expand();
    }
    /**
     * @return {?}
     */
    onCloseClick() {
        this.close.emit();
    }
    /**
     * @return {?}
     */
    onToggleFullScreenClick() {
        this.fullscreen.emit();
    }
}
WidgetControlsComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widget-controls',
                template: `<div>
  <div class='title'>{{widget.title}}</div>
  <div *ngIf='!widget.isFullScreen; else fullscreen' class='buttons no-highlighting not-grabbable'>
    <span *ngIf='widget.isExpanded; else expand' class='material-icons' (click)='onCollapseClick()'>expand_less</span>
    <ng-template #expand>
      <span class='material-icons' (click)='onExpandClick()'>expand_more</span>
    </ng-template>
    <span class='material-icons d-none d-sm-inline' (click)='onToggleFullScreenClick()'>fullscreen</span>
    <span class='material-icons' (click)='onCloseClick()'>close</span>
  </div>
  <ng-template #fullscreen>
    <div class='material-icons buttons no-highlighting not-grabbable' (click)='onToggleFullScreenClick()'>fullscreen_exit</div>
  </ng-template>
</div>`,
                styles: [`:host{display:block;cursor:-webkit-grab;cursor:grab;overflow:hidden;margin-bottom:15px}:host>div>.title{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host>div>div:first-of-type{float:left}:host>div>div:last-of-type{float:right;cursor:default}.fullscreen :host{cursor:default}`]
            },] },
];
/** @nocollapse */
WidgetControlsComponent.ctorParameters = () => [
    { type: ElementRef, },
];
WidgetControlsComponent.propDecorators = {
    "widget": [{ type: Input },],
    "close": [{ type: Output },],
    "fullscreen": [{ type: Output },],
};
function WidgetControlsComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    WidgetControlsComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    WidgetControlsComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    WidgetControlsComponent.propDecorators;
    /** @type {?} */
    WidgetControlsComponent.prototype.widget;
    /** @type {?} */
    WidgetControlsComponent.prototype.close;
    /** @type {?} */
    WidgetControlsComponent.prototype.fullscreen;
    /** @type {?} */
    WidgetControlsComponent.prototype.elementRef;
}
