// @navihealth/architect-list-widget/bundles/navihealth-architect-list-widget.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory(require('@angular/core'), require('@navihealth/browser-bus'), require('@angular/platform-browser'), require('@angular/platform-browser-dynamic')) :
    typeof define === 'function' && define.amd ? define('@navihealth/architect-list-widget', ['@angular/core', '@navihealth/browser-bus', '@angular/platform-browser', '@angular/platform-browser-dynamic'], factory) :
    (global.navihealth = global.navihealth || {}, global.navihealth['architect-list-widget'] = factory(global.ng.core,null,global.ng.platformBrowser,global.ng.platformBrowserDynamic));
}(this, (function (core,browserBus,platformBrowser,platformBrowserDynamic) { 'use strict';

    /*! *****************************************************************************
    Copyright (c) Microsoft Corporation. All rights reserved.
    Licensed under the Apache License, Version 2.0 (the "License"); you may not use
    this file except in compliance with the License. You may obtain a copy of the
    License at http://www.apache.org/licenses/LICENSE-2.0

    THIS CODE IS PROVIDED ON AN *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
    WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
    MERCHANTABLITY OR NON-INFRINGEMENT.

    See the Apache Version 2.0 License for specific language governing permissions
    and limitations under the License.
    ***************************************************************************** */
    function __awaiter(thisArg, _arguments, P, generator) {
        return new (P || (P = Promise))(function (resolve, reject) {
            function fulfilled(value) { try {
                step(generator.next(value));
            }
            catch (e) {
                reject(e);
            } }
            function rejected(value) { try {
                step(generator["throw"](value));
            }
            catch (e) {
                reject(e);
            } }
            function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
            step((generator = generator.apply(thisArg, _arguments || [])).next());
        });
    }
    function __generator(thisArg, body) {
        var _ = { label: 0, sent: function () { if (t[0] & 1)
                throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
        return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function () { return this; }), g;
        function verb(n) { return function (v) { return step([n, v]); }; }
        function step(op) {
            if (f)
                throw new TypeError("Generator is already executing.");
            while (_)
                try {
                    if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done)
                        return t;
                    if (y = 0, t)
                        op = [op[0] & 2, t.value];
                    switch (op[0]) {
                        case 0:
                        case 1:
                            t = op;
                            break;
                        case 4:
                            _.label++;
                            return { value: op[1], done: false };
                        case 5:
                            _.label++;
                            y = op[1];
                            op = [0];
                            continue;
                        case 7:
                            op = _.ops.pop();
                            _.trys.pop();
                            continue;
                        default:
                            if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
                                _ = 0;
                                continue;
                            }
                            if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) {
                                _.label = op[1];
                                break;
                            }
                            if (op[0] === 6 && _.label < t[1]) {
                                _.label = t[1];
                                t = op;
                                break;
                            }
                            if (t && _.label < t[2]) {
                                _.label = t[2];
                                _.ops.push(op);
                                break;
                            }
                            if (t[2])
                                _.ops.pop();
                            _.trys.pop();
                            continue;
                    }
                    op = body.call(thisArg, _);
                }
                catch (e) {
                    op = [6, e];
                    y = 0;
                }
                finally {
                    f = t = 0;
                }
            if (op[0] & 5)
                throw op[1];
            return { value: op[0] ? op[1] : void 0, done: true };
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var AppComponent = (function () {
        function AppComponent(elementRef) {
            this.elementRef = elementRef;
            this.title = 'architects';
            this.architects = [
                'Seshan',
                'Aneesh',
                'Nijil'
            ];
            this.bus = ((browserBus.BusFactory)).create();
        }
        /**
         * @param {?} event
         * @return {?}
         */
        AppComponent.prototype.newArchitectOnKeyUp = /**
         * @param {?} event
         * @return {?}
         */
            function (event) {
                this.newArchitectName = event.target.value;
            };
        /**
         * @return {?}
         */
        AppComponent.prototype.addOnClick = /**
         * @return {?}
         */
            function () {
                if (this.newArchitectName) {
                    this.architects.push(this.newArchitectName);
                    this.notifyArchitectSelected(this.newArchitectName);
                }
            };
        /**
         * @param {?} architect
         * @return {?}
         */
        AppComponent.prototype.architectOnClick = /**
         * @param {?} architect
         * @return {?}
         */
            function (architect) {
                this.notifyArchitectSelected(architect);
            };
        /**
         * @return {?}
         */
        AppComponent.prototype.ngAfterContentInit = /**
         * @return {?}
         */
            function () {
                this.elementRef.nativeElement.classList.add('architect-list-widget');
            };
        /**
         * @param {?} architect
         * @return {?}
         */
        AppComponent.prototype.notifyArchitectSelected = /**
         * @param {?} architect
         * @return {?}
         */
            function (architect) {
                this.bus.notify({
                    eventName: 'architect-selected',
                    message: {
                        name: architect
                    }
                });
            };
        AppComponent.decorators = [
            { type: core.Component, args: [{
                        selector: 'app-architect-list-widget',
                        template: "<div class='col-xs-12'>\n  <input (keyup)='newArchitectOnKeyUp($event)' placeholder='enter name' />\n  <button (click)='addOnClick()' class='btn btn-primary'>Add Architect</button>\n  <ul>\n    <li *ngFor='let architect of architects' (click)='architectOnClick(architect)'>{{ architect }}</li>\n  </ul>\n</div>",
                        styles: [":host.architect-list-widget ul{margin-top:10px;padding:0}:host.architect-list-widget li{list-style:none;color:green}"]
                    },] },
        ];
        /** @nocollapse */
        AppComponent.ctorParameters = function () {
            return [
                { type: core.ElementRef, },
            ];
        };
        return AppComponent;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var AppModule = (function () {
        function AppModule() {
        }
        AppModule.decorators = [
            { type: core.NgModule, args: [{
                        declarations: [
                            AppComponent
                        ],
                        imports: [
                            platformBrowser.BrowserModule
                        ],
                        providers: [],
                        bootstrap: [AppComponent]
                    },] },
        ];
        return AppModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var /** @type {?} */ id = 'd4181c23-4c1f-4c2a-910a-7d383afb9cd9';
    var /** @type {?} */ title = 'Architect List (Angular 5)';
    var /** @type {?} */ width = {
        xs: 12,
        sm: 8,
        md: 6,
        lg: 5,
        xl: 4
    };
    var /** @type {?} */ initialHeight = '250px';
    var /** @type {?} */ render = function (container) {
        return __awaiter(this, void 0, void 0, function () {
            var _this = this;
            return __generator(this, function (_a) {
                return [2 /*return*/, new Promise(function (resolve, reject) {
                        return __awaiter(_this, void 0, void 0, function () {
                            var appRoot, error_1;
                            return __generator(this, function (_a) {
                                switch (_a.label) {
                                    case 0:
                                        _a.trys.push([0, 2, , 3]);
                                        appRoot = document.createElement('app-architect-list-widget');
                                        container.appendChild(appRoot);
                                        return [4 /*yield*/, platformBrowserDynamic.platformBrowserDynamic().bootstrapModule(AppModule)];
                                    case 1:
                                        _a.sent();
                                        resolve();
                                        return [3 /*break*/, 3];
                                    case 2:
                                        error_1 = _a.sent();
                                        reject(error_1);
                                        return [3 /*break*/, 3];
                                    case 3: return [2 /*return*/];
                                }
                            });
                        });
                    })];
            });
        });
    };
    var widget = {
        id: id,
        title: title,
        width: width,
        initialHeight: initialHeight,
        render: render
    };

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    return widget;

})));
