// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/bootstrap.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Injectable } from '@angular/core';
var BootstrapService = /** @class */ (function () {
    function BootstrapService() {
        this.maxColumns = 12;
        this.columnPrefix = 'col-';
        this.breakpoints = {
            xl: 'd-xl-none',
            lg: 'd-lg-none',
            md: 'd-md-none',
            sm: 'd-sm-none',
            xs: 'd-none'
        };
    }
    /**
     * @return {?}
     */
    BootstrapService.prototype.getCurrentBreakpoint = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ breakpoint = '';
        var /** @type {?} */ marker = document.createElement('div');
        marker.style.visibility = 'hidden';
        document.body.appendChild(marker);
        var /** @type {?} */ breakpointKeys = Object.keys(this.breakpoints);
        for (var /** @type {?} */ i = 0; i < breakpointKeys.length; i++) {
            breakpoint = breakpointKeys[i];
            marker.classList.add(this.breakpoints[breakpoint]);
            var /** @type {?} */ style = window.getComputedStyle(marker);
            if (style.display === 'none') {
                break;
            }
        }
        marker.remove();
        return breakpoint;
    };
    /**
     * @return {?}
     */
    BootstrapService.prototype.hasBreakpointChanged = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ breakpoint = this.getCurrentBreakpoint();
        if (breakpoint !== this.currentBreakpoint) {
            this.currentBreakpoint = breakpoint;
            return true;
        }
        return false;
    };
    /**
     * @param {?} getWidth
     * @return {?}
     */
    BootstrapService.prototype.getWidth = /**
     * @param {?} getWidth
     * @return {?}
     */
    function (getWidth) {
        var /** @type {?} */ breakpoint = this.getCurrentBreakpoint();
        var /** @type {?} */ width = getWidth(breakpoint);
        if (!width) {
            // Look for the largest breakpoint with a width that is smaller than the current breakpoint.
            var /** @type {?} */ smallerBreakpoints = this.getSmallerBreakpoints(breakpoint);
            while (!width && smallerBreakpoints.length) {
                breakpoint = smallerBreakpoints.shift();
                width = getWidth(breakpoint);
            }
        }
        return width;
    };
    /**
     * @param {?} makeFullWidth
     * @return {?}
     */
    BootstrapService.prototype.makeFullWidth = /**
     * @param {?} makeFullWidth
     * @return {?}
     */
    function (makeFullWidth) {
        var _this = this;
        var /** @type {?} */ breakpoint = this.getCurrentBreakpoint();
        // Make each breakpoint at the current size and smaller the full width of a row.
        this.getSmallerBreakpoints(breakpoint)
            .concat([breakpoint])
            .forEach(function (x) {
            makeFullWidth(x, _this.maxColumns);
        });
    };
    /**
     * @param {?} widget
     * @param {?} targetWidthOfWidget
     * @return {?}
     */
    BootstrapService.prototype.resize = /**
     * @param {?} widget
     * @param {?} targetWidthOfWidget
     * @return {?}
     */
    function (widget, targetWidthOfWidget) {
        var /** @type {?} */ breakpoint = this.getCurrentBreakpoint();
        var /** @type {?} */ width = widget.getCurrentWidth(breakpoint);
        if (!width) {
            // Look for the largest breakpoint with a width that is smaller than the current breakpoint.
            var /** @type {?} */ smallerBreakpoints = this.getSmallerBreakpoints(breakpoint);
            while (!width && smallerBreakpoints.length) {
                breakpoint = smallerBreakpoints.shift();
                width = widget.getCurrentWidth(breakpoint);
            }
        }
        widget.resize(breakpoint, targetWidthOfWidget);
    };
    /**
     * @param {?} widget
     * @return {?}
     */
    BootstrapService.prototype.getClasses = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        var /** @type {?} */ widths = widget.getCurrentWidths();
        return Object.keys(widths)
            .map(function (x) {
            var /** @type {?} */ width = widths[x];
            if (x === 'xs') {
                return 'col-' + width;
            }
            return 'col-' + x + '-' + width;
        });
    };
    /**
     * @param {?} breakpoint
     * @return {?}
     */
    BootstrapService.prototype.getSmallerBreakpoints = /**
     * @param {?} breakpoint
     * @return {?}
     */
    function (breakpoint) {
        var /** @type {?} */ allBreakpoints = Object.keys(this.breakpoints);
        var /** @type {?} */ smallerBreakpoints = allBreakpoints.slice(allBreakpoints.indexOf(breakpoint) + 1, allBreakpoints.length);
        return smallerBreakpoints;
    };
    BootstrapService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    BootstrapService.ctorParameters = function () { return []; };
    return BootstrapService;
}());
export { BootstrapService };
function BootstrapService_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    BootstrapService.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    BootstrapService.ctorParameters;
    /** @type {?} */
    BootstrapService.prototype.maxColumns;
    /** @type {?} */
    BootstrapService.prototype.columnPrefix;
    /** @type {?} */
    BootstrapService.prototype.breakpoints;
    /** @type {?} */
    BootstrapService.prototype.currentBreakpoint;
}
