// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/layout-editor.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
var LayoutEditor = /** @class */ (function () {
    function LayoutEditor() {
        this.isDisplayed = false;
    }
    /**
     * @param {?} xPosition
     * @param {?} yPosition
     * @return {?}
     */
    LayoutEditor.prototype.edit = /**
     * @param {?} xPosition
     * @param {?} yPosition
     * @return {?}
     */
    function (xPosition, yPosition) {
        this.isDisplayed = true;
        this.xPosition = xPosition;
        this.yPosition = yPosition;
    };
    /**
     * @return {?}
     */
    LayoutEditor.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.isDisplayed = false;
        this.xPosition = null;
        this.yPosition = null;
    };
    return LayoutEditor;
}());
export { LayoutEditor };
function LayoutEditor_tsickle_Closure_declarations() {
    /** @type {?} */
    LayoutEditor.prototype.isDisplayed;
    /** @type {?} */
    LayoutEditor.prototype.xPosition;
    /** @type {?} */
    LayoutEditor.prototype.yPosition;
}
