// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/library.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
/**
 * @record
 */
export function ILibraryWidget() { }
function ILibraryWidget_tsickle_Closure_declarations() {
    /** @type {?} */
    ILibraryWidget.prototype.isRendered;
    /** @type {?} */
    ILibraryWidget.prototype.widget;
}
var Library = /** @class */ (function () {
    function Library(renderedWidgets, availableWidgets) {
        this.renderedWidgets = renderedWidgets;
        this.availableWidgets = availableWidgets;
        this.isDisplayed = false;
        this.buildLibrary();
    }
    /**
     * @param {?} widget
     * @return {?}
     */
    Library.prototype.isWidgetRendered = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        return this.renderedWidgets
            .map(function (renderedWidget) { return renderedWidget.id; })
            .indexOf(widget.id) >= 0;
    };
    /**
     * @return {?}
     */
    Library.prototype.buildLibrary = /**
     * @return {?}
     */
    function () {
        var _this = this;
        this.widgets = this.availableWidgets.map(function (widget) {
            var /** @type {?} */ isRendered = _this.isWidgetRendered(widget);
            return { isRendered: isRendered, widget: widget };
        });
    };
    /**
     * @param {?} xPosition
     * @param {?} yPosition
     * @return {?}
     */
    Library.prototype.edit = /**
     * @param {?} xPosition
     * @param {?} yPosition
     * @return {?}
     */
    function (xPosition, yPosition) {
        this.isDisplayed = true;
        this.xPosition = xPosition;
        this.yPosition = yPosition;
    };
    /**
     * @return {?}
     */
    Library.prototype.hide = /**
     * @return {?}
     */
    function () {
        this.isDisplayed = false;
        this.xPosition = null;
        this.yPosition = null;
    };
    /**
     * @param {?} widget
     * @return {?}
     */
    Library.prototype.renderWidget = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        this.renderedWidgets.push(widget);
        this.buildLibrary();
    };
    /**
     * @param {?} widget
     * @return {?}
     */
    Library.prototype.derenderWidget = /**
     * @param {?} widget
     * @return {?}
     */
    function (widget) {
        var _this = this;
        this.renderedWidgets
            .filter(function (x) { return x.id === widget.id; })
            .forEach(function (x) {
            _this.renderedWidgets.splice(_this.renderedWidgets.indexOf(x), 1);
        });
        this.buildLibrary();
    };
    return Library;
}());
export { Library };
function Library_tsickle_Closure_declarations() {
    /** @type {?} */
    Library.prototype.isDisplayed;
    /** @type {?} */
    Library.prototype.xPosition;
    /** @type {?} */
    Library.prototype.yPosition;
    /** @type {?} */
    Library.prototype.widgets;
    /** @type {?} */
    Library.prototype.renderedWidgets;
    /** @type {?} */
    Library.prototype.availableWidgets;
}
