// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/spacer/spacer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, Output, EventEmitter } from '@angular/core';
export class SpacerComponent {
    /**
     * @param {?} elementRef
     */
    constructor(elementRef) {
        this.elementRef = elementRef;
        this.spacerHovered = new EventEmitter();
        this.spacerUnhovered = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.elementRef.nativeElement.classList.add('col-12');
    }
    /**
     * @return {?}
     */
    onMouseEnter() {
        this.spacerHovered.emit(this);
    }
    /**
     * @return {?}
     */
    onTouchEnter() {
        this.spacerHovered.emit(this);
    }
    /**
     * @return {?}
     */
    onMouseLeave() {
        this.spacerUnhovered.emit();
    }
    /**
     * @return {?}
     */
    onTouchLeave() {
        this.spacerUnhovered.emit();
    }
}
SpacerComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widgets-panel-spacer',
                template: `<div (mouseenter)='onMouseEnter()' (touchenter)='onTouchEnter()' (mouseleave)='onMouseLeave()' (touchleave)='onTouchLeave()'>
</div>`,
                styles: [`:host{height:30px}:host>div{height:100%}`]
            },] },
];
/** @nocollapse */
SpacerComponent.ctorParameters = () => [
    { type: ElementRef, },
];
SpacerComponent.propDecorators = {
    "spacer": [{ type: Input },],
    "spacerHovered": [{ type: Output },],
    "spacerUnhovered": [{ type: Output },],
};
function SpacerComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SpacerComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SpacerComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SpacerComponent.propDecorators;
    /** @type {?} */
    SpacerComponent.prototype.spacer;
    /** @type {?} */
    SpacerComponent.prototype.spacerHovered;
    /** @type {?} */
    SpacerComponent.prototype.spacerUnhovered;
    /** @type {?} */
    SpacerComponent.prototype.elementRef;
}
