// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/layout-editor/layout-editor.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { LayoutEditor } from '../layout-editor.model';
var LayoutEditorComponent = /** @class */ (function () {
    function LayoutEditorComponent(elementRef) {
        this.elementRef = elementRef;
        this.done = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LayoutEditorComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    LayoutEditorComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.classList.add('card', 'card-row', 'col-8', 'col-sm-6', 'popup');
    };
    /**
     * @return {?}
     */
    LayoutEditorComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ width = this.elementRef.nativeElement
            .getBoundingClientRect()
            .width;
        this.elementRef.nativeElement.style.top = (this.layoutEditor.yPosition - 20) + 'px';
        this.elementRef.nativeElement.style.left = (this.layoutEditor.xPosition - width + 20) + 'px';
    };
    /**
     * @return {?}
     */
    LayoutEditorComponent.prototype.onDoneClick = /**
     * @return {?}
     */
    function () {
        this.done.emit();
    };
    LayoutEditorComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-widgets-layout-editor',
                    template: "<div class='title'>Layout Editor</div>\n<p>\n  <b>Coming Soon!</b>\n</p>\n<p>\n  This will allow some users to edit pages and make custom layouts targeted for specific roles in your organization.\n</p>\n<p>\n  For now, changes made will not be saved.\n</p>\n<div class='button-container'>\n  <button class='btn btn-primary' (click)='onDoneClick()'>Done</button>\n</div>",
                    styles: [":host{position:absolute;z-index:100;background:#fff}"]
                },] },
    ];
    /** @nocollapse */
    LayoutEditorComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    LayoutEditorComponent.propDecorators = {
        "layoutEditor": [{ type: Input },],
        "done": [{ type: Output },],
    };
    return LayoutEditorComponent;
}());
export { LayoutEditorComponent };
function LayoutEditorComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    LayoutEditorComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    LayoutEditorComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    LayoutEditorComponent.propDecorators;
    /** @type {?} */
    LayoutEditorComponent.prototype.layoutEditor;
    /** @type {?} */
    LayoutEditorComponent.prototype.done;
    /** @type {?} */
    LayoutEditorComponent.prototype.elementRef;
}
