// @navihealth/architect-list-widget/esm5/widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { enableProdMode } from '@angular/core';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppModule } from './app/app.module';
import { environment } from './environments/environment';
if (environment.production) {
    enableProdMode();
}
var /** @type {?} */ id = 'd4181c23-4c1f-4c2a-910a-7d383afb9cd9';
var /** @type {?} */ title = 'Architect List (Angular 5)';
var /** @type {?} */ width = {
    xs: 12,
    sm: 8,
    md: 6,
    lg: 5,
    xl: 4
};
var /** @type {?} */ initialHeight = '250px';
var /** @type {?} */ render = function (container) {
    return tslib_1.__awaiter(this, void 0, void 0, function () {
        var _this = this;
        return tslib_1.__generator(this, function (_a) {
            return [2 /*return*/, new Promise(function (resolve, reject) { return tslib_1.__awaiter(_this, void 0, void 0, function () {
                    var appRoot, error_1;
                    return tslib_1.__generator(this, function (_a) {
                        switch (_a.label) {
                            case 0:
                                _a.trys.push([0, 2, , 3]);
                                appRoot = document.createElement('app-architect-list-widget');
                                container.appendChild(appRoot);
                                return [4 /*yield*/, platformBrowserDynamic().bootstrapModule(AppModule)];
                            case 1:
                                _a.sent();
                                resolve();
                                return [3 /*break*/, 3];
                            case 2:
                                error_1 = _a.sent();
                                reject(error_1);
                                return [3 /*break*/, 3];
                            case 3: return [2 /*return*/];
                        }
                    });
                }); })];
        });
    });
};
var ɵ0 = render;
export default {
    id: id,
    title: title,
    width: width,
    initialHeight: initialHeight,
    render: render
};
export { ɵ0 };
