// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/library/library.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, ElementRef, EventEmitter } from '@angular/core';
import { Library } from '../library.model';
var LibraryComponent = /** @class */ (function () {
    function LibraryComponent(elementRef) {
        this.elementRef = elementRef;
        this.done = new EventEmitter();
        this.renderWidget = new EventEmitter();
        this.derenderWidget = new EventEmitter();
    }
    /**
     * @return {?}
     */
    LibraryComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    LibraryComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.classList.add('card', 'card-row', 'popup');
    };
    /**
     * @return {?}
     */
    LibraryComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ width = this.elementRef.nativeElement
            .getBoundingClientRect()
            .width;
        this.elementRef.nativeElement.style.top = (this.library.yPosition - 20) + 'px';
        this.elementRef.nativeElement.style.left = (this.library.xPosition - width + 20) + 'px';
    };
    /**
     * @return {?}
     */
    LibraryComponent.prototype.onDoneClick = /**
     * @return {?}
     */
    function () {
        this.done.emit();
    };
    /**
     * @param {?} libraryWidget
     * @return {?}
     */
    LibraryComponent.prototype.onCheckboxChange = /**
     * @param {?} libraryWidget
     * @return {?}
     */
    function (libraryWidget) {
        if (libraryWidget.isRendered) {
            this.library.derenderWidget(libraryWidget.widget);
            this.derenderWidget.emit(libraryWidget.widget);
        }
        else {
            this.library.renderWidget(libraryWidget.widget);
            this.renderWidget.emit(libraryWidget.widget);
        }
    };
    LibraryComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-widgets-library',
                    template: "<div class='title'>Library</div>\n<ul class='list-group'>\n  <li *ngFor='let libraryWidget of library.widgets' class='list-group-item'>\n    <span>{{libraryWidget.widget.title}}</span>\n    <!-- <span *ngIf='libraryWidget.isRendered; else not' class='material-icons' (click)='onRemoveWidget(libraryWidget.widget)'>remove_circle_outline</span>\n    <ng-template #not>\n      <span class='material-icons' (click)='onAddWidget(libraryWidget.widget)'>add_circle_outline</span>\n    </ng-template> -->\n    <!-- <div class='pretty p-switch p-fill'>\n      <input id='libraryWidget.widget.id' type='checkbox' [checked]='libraryWidget.isRendered' (change)='onCheckboxChange(libraryWidget)' />\n    </div> -->\n    <span class='pretty p-switch p-fill'>\n      <input type='checkbox' [checked]='libraryWidget.isRendered' (change)='onCheckboxChange(libraryWidget)' />\n      <div class='state p-success'>\n        <label></label>\n      </div>\n    </span>\n  </li>\n</ul>\n<div class='button-container'>\n  <button class='btn btn-primary' (click)='onDoneClick()'>Done</button>\n</div>",
                    styles: [":host{position:absolute;z-index:100;background:#fff}ul{list-style:none;padding:0;margin-bottom:20px}ul>li{overflow:hidden;text-overflow:ellipsis}ul>li>span:first-of-type{float:left;margin-right:30px}ul>li>span:last-of-type{float:right;margin-top:5px}label:after{margin-top:-1px}.pretty input{width:200%}"]
                },] },
    ];
    /** @nocollapse */
    LibraryComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    LibraryComponent.propDecorators = {
        "library": [{ type: Input },],
        "done": [{ type: Output },],
        "renderWidget": [{ type: Output },],
        "derenderWidget": [{ type: Output },],
    };
    return LibraryComponent;
}());
export { LibraryComponent };
function LibraryComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    LibraryComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    LibraryComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    LibraryComponent.propDecorators;
    /** @type {?} */
    LibraryComponent.prototype.library;
    /** @type {?} */
    LibraryComponent.prototype.done;
    /** @type {?} */
    LibraryComponent.prototype.renderWidget;
    /** @type {?} */
    LibraryComponent.prototype.derenderWidget;
    /** @type {?} */
    LibraryComponent.prototype.elementRef;
}
