// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/spacer/spacer.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, ElementRef, Input, Output, EventEmitter } from '@angular/core';
var SpacerComponent = /** @class */ (function () {
    function SpacerComponent(elementRef) {
        this.elementRef = elementRef;
        this.spacerHovered = new EventEmitter();
        this.spacerUnhovered = new EventEmitter();
    }
    /**
     * @return {?}
     */
    SpacerComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    SpacerComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.classList.add('col-12');
    };
    /**
     * @return {?}
     */
    SpacerComponent.prototype.onMouseEnter = /**
     * @return {?}
     */
    function () {
        this.spacerHovered.emit(this);
    };
    /**
     * @return {?}
     */
    SpacerComponent.prototype.onTouchEnter = /**
     * @return {?}
     */
    function () {
        this.spacerHovered.emit(this);
    };
    /**
     * @return {?}
     */
    SpacerComponent.prototype.onMouseLeave = /**
     * @return {?}
     */
    function () {
        this.spacerUnhovered.emit();
    };
    /**
     * @return {?}
     */
    SpacerComponent.prototype.onTouchLeave = /**
     * @return {?}
     */
    function () {
        this.spacerUnhovered.emit();
    };
    SpacerComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-widgets-panel-spacer',
                    template: "<div (mouseenter)='onMouseEnter()' (touchenter)='onTouchEnter()' (mouseleave)='onMouseLeave()' (touchleave)='onTouchLeave()'>\n</div>",
                    styles: [":host{height:30px}:host>div{height:100%}"]
                },] },
    ];
    /** @nocollapse */
    SpacerComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    SpacerComponent.propDecorators = {
        "spacer": [{ type: Input },],
        "spacerHovered": [{ type: Output },],
        "spacerUnhovered": [{ type: Output },],
    };
    return SpacerComponent;
}());
export { SpacerComponent };
function SpacerComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    SpacerComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    SpacerComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    SpacerComponent.propDecorators;
    /** @type {?} */
    SpacerComponent.prototype.spacer;
    /** @type {?} */
    SpacerComponent.prototype.spacerHovered;
    /** @type {?} */
    SpacerComponent.prototype.spacerUnhovered;
    /** @type {?} */
    SpacerComponent.prototype.elementRef;
}
