// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/widgets/resize-guideline/resize-guideline.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { BootstrapService } from '../bootstrap.service';
export class ResizeGuidelineComponent {
    /**
     * @param {?} elementRef
     * @param {?} bootstrapService
     */
    constructor(elementRef, bootstrapService) {
        this.elementRef = elementRef;
        this.bootstrapService = bootstrapService;
        this.resized = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() { }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        const /** @type {?} */ thisElement = this.elementRef.nativeElement;
        thisElement.classList.add(this.bootstrapService.columnPrefix + this.guideline.widthOfGuideline);
        thisElement.style.top = (this.guideline.top + window.scrollY) + 'px';
        thisElement.style.height = this.guideline.height + 'px';
    }
    /**
     * @return {?}
     */
    onHover() {
        const /** @type {?} */ widget = this.guideline.widget;
        if (this.guideline.targetWidthOfWidget !== this.bootstrapService.getWidth(widget.getCurrentWidth.bind(widget))) {
            this.bootstrapService.resize(this.guideline.widget, this.guideline.targetWidthOfWidget);
            this.resized.emit(this.guideline.widget);
        }
    }
}
ResizeGuidelineComponent.decorators = [
    { type: Component, args: [{
                selector: 'app-widget-resize-guideline',
                template: `<div (mouseenter)='onHover()' (touchenter)='onHover()'></div>`,
                styles: [`:host{position:absolute}:host>div{border-right:1px dotted green;height:100%}`]
            },] },
];
/** @nocollapse */
ResizeGuidelineComponent.ctorParameters = () => [
    { type: ElementRef, },
    { type: BootstrapService, },
];
ResizeGuidelineComponent.propDecorators = {
    "guideline": [{ type: Input },],
    "resized": [{ type: Output },],
};
function ResizeGuidelineComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ResizeGuidelineComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ResizeGuidelineComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ResizeGuidelineComponent.propDecorators;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.guideline;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.resized;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.elementRef;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.bootstrapService;
}
