// @navihealth/architect-list-widget/esm2015/widget.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { enableProdMode } from '@angular/core';
import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
import { AppModule } from './app/app.module';
import { environment } from './environments/environment';
if (environment.production) {
    enableProdMode();
}
const /** @type {?} */ id = 'd4181c23-4c1f-4c2a-910a-7d383afb9cd9';
const /** @type {?} */ title = 'Architect List (Angular 5)';
const /** @type {?} */ width = {
    xs: 12,
    sm: 8,
    md: 6,
    lg: 5,
    xl: 4
};
const /** @type {?} */ initialHeight = '250px';
const /** @type {?} */ render = function (container) {
    return tslib_1.__awaiter(this, void 0, void 0, function* () {
        return new Promise((resolve, reject) => tslib_1.__awaiter(this, void 0, void 0, function* () {
            try {
                const /** @type {?} */ appRoot = document.createElement('app-architect-list-widget');
                container.appendChild(appRoot);
                yield platformBrowserDynamic().bootstrapModule(AppModule);
                resolve();
            }
            catch (/** @type {?} */ error) {
                reject(error);
            }
        }));
    });
};
const ɵ0 = render;
export default {
    id,
    title,
    width,
    initialHeight,
    render
};
export { ɵ0 };
