// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/widget-controls/widget-controls.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { Widget } from '../widget.model';
var WidgetControlsComponent = /** @class */ (function () {
    function WidgetControlsComponent(elementRef) {
        this.elementRef = elementRef;
        this.close = new EventEmitter();
        this.fullscreen = new EventEmitter();
    }
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.elementRef.nativeElement.classList.add('col-12');
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.ngAfterViewInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        // We'll set the width of the title based on the width left by the buttons.
        // This will make the overflow on the title work properly. But we won't
        // do this until the stack is empty. Without the setTime, the first page load
        // after a Hard Reload does not calculate the offsetWidths correctly.
        setTimeout(function () {
            var /** @type {?} */ containerElement = _this.elementRef.nativeElement.children[0];
            var /** @type {?} */ controlsWidth = containerElement.offsetWidth;
            var /** @type {?} */ buttonsElement = containerElement.querySelector('.buttons');
            var /** @type {?} */ buttonsWidth = buttonsElement.offsetWidth;
            var /** @type {?} */ padding = 8;
            var /** @type {?} */ titleElement = containerElement.querySelector('.title');
            titleElement.style.width = (100 - padding - (buttonsWidth / controlsWidth * 100)) + '%';
        }, 0);
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.onCollapseClick = /**
     * @return {?}
     */
    function () {
        this.widget.collapse();
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.onExpandClick = /**
     * @return {?}
     */
    function () {
        this.widget.expand();
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.onCloseClick = /**
     * @return {?}
     */
    function () {
        this.close.emit();
    };
    /**
     * @return {?}
     */
    WidgetControlsComponent.prototype.onToggleFullScreenClick = /**
     * @return {?}
     */
    function () {
        this.fullscreen.emit();
    };
    WidgetControlsComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-widget-controls',
                    template: "<div>\n  <div class='title'>{{widget.title}}</div>\n  <div *ngIf='!widget.isFullScreen; else fullscreen' class='buttons no-highlighting not-grabbable'>\n    <span *ngIf='widget.isExpanded; else expand' class='material-icons' (click)='onCollapseClick()'>expand_less</span>\n    <ng-template #expand>\n      <span class='material-icons' (click)='onExpandClick()'>expand_more</span>\n    </ng-template>\n    <span class='material-icons d-none d-sm-inline' (click)='onToggleFullScreenClick()'>fullscreen</span>\n    <span class='material-icons' (click)='onCloseClick()'>close</span>\n  </div>\n  <ng-template #fullscreen>\n    <div class='material-icons buttons no-highlighting not-grabbable' (click)='onToggleFullScreenClick()'>fullscreen_exit</div>\n  </ng-template>\n</div>",
                    styles: [":host{display:block;cursor:-webkit-grab;cursor:grab;overflow:hidden;margin-bottom:15px}:host>div>.title{text-overflow:ellipsis;white-space:nowrap;overflow:hidden}:host>div>div:first-of-type{float:left}:host>div>div:last-of-type{float:right;cursor:default}.fullscreen :host{cursor:default}"]
                },] },
    ];
    /** @nocollapse */
    WidgetControlsComponent.ctorParameters = function () { return [
        { type: ElementRef, },
    ]; };
    WidgetControlsComponent.propDecorators = {
        "widget": [{ type: Input },],
        "close": [{ type: Output },],
        "fullscreen": [{ type: Output },],
    };
    return WidgetControlsComponent;
}());
export { WidgetControlsComponent };
function WidgetControlsComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    WidgetControlsComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    WidgetControlsComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    WidgetControlsComponent.propDecorators;
    /** @type {?} */
    WidgetControlsComponent.prototype.widget;
    /** @type {?} */
    WidgetControlsComponent.prototype.close;
    /** @type {?} */
    WidgetControlsComponent.prototype.fullscreen;
    /** @type {?} */
    WidgetControlsComponent.prototype.elementRef;
}
