// @navihealth/widget-dashboard-ng5/esm5/src/dashboard/widgets/resize-guideline/resize-guideline.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input, Output, EventEmitter, ElementRef } from '@angular/core';
import { BootstrapService } from '../bootstrap.service';
var ResizeGuidelineComponent = /** @class */ (function () {
    function ResizeGuidelineComponent(elementRef, bootstrapService) {
        this.elementRef = elementRef;
        this.bootstrapService = bootstrapService;
        this.resized = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ResizeGuidelineComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () { };
    /**
     * @return {?}
     */
    ResizeGuidelineComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ thisElement = this.elementRef.nativeElement;
        thisElement.classList.add(this.bootstrapService.columnPrefix + this.guideline.widthOfGuideline);
        thisElement.style.top = (this.guideline.top + window.scrollY) + 'px';
        thisElement.style.height = this.guideline.height + 'px';
    };
    /**
     * @return {?}
     */
    ResizeGuidelineComponent.prototype.onHover = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ widget = this.guideline.widget;
        if (this.guideline.targetWidthOfWidget !== this.bootstrapService.getWidth(widget.getCurrentWidth.bind(widget))) {
            this.bootstrapService.resize(this.guideline.widget, this.guideline.targetWidthOfWidget);
            this.resized.emit(this.guideline.widget);
        }
    };
    ResizeGuidelineComponent.decorators = [
        { type: Component, args: [{
                    selector: 'app-widget-resize-guideline',
                    template: "<div (mouseenter)='onHover()' (touchenter)='onHover()'></div>",
                    styles: [":host{position:absolute}:host>div{border-right:1px dotted green;height:100%}"]
                },] },
    ];
    /** @nocollapse */
    ResizeGuidelineComponent.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: BootstrapService, },
    ]; };
    ResizeGuidelineComponent.propDecorators = {
        "guideline": [{ type: Input },],
        "resized": [{ type: Output },],
    };
    return ResizeGuidelineComponent;
}());
export { ResizeGuidelineComponent };
function ResizeGuidelineComponent_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ResizeGuidelineComponent.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ResizeGuidelineComponent.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ResizeGuidelineComponent.propDecorators;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.guideline;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.resized;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.elementRef;
    /** @type {?} */
    ResizeGuidelineComponent.prototype.bootstrapService;
}
