// @navihealth/widget-dashboard-ng5/esm2015/src/dashboard/dashboard.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { BootstrapService } from './widgets/bootstrap.service';
import { DashboardComponent } from './dashboard.component';
import { PanelComponent } from './widgets/panel/panel.component';
import { PanelControlsComponent } from './widgets/panel-controls/panel-controls.component';
import { SpacerComponent } from './widgets/spacer/spacer.component';
import { WidgetComponent } from './widgets/widget/widget.component';
import { WidgetControlsComponent } from './widgets/widget-controls/widget-controls.component';
import { LibraryComponent } from './widgets/library/library.component';
import { DropTargetComponent } from './widgets/drop-target/drop-target.component';
import { ResizeGuidelineComponent } from './widgets/resize-guideline/resize-guideline.component';
import { LayoutEditorComponent } from './widgets/layout-editor/layout-editor.component';
export class DashboardModule {
    /**
     * @param {?} WidgetsService
     * @return {?}
     */
    static forRoot(WidgetsService) {
        return {
            ngModule: DashboardModule,
            providers: [
                { provide: 'widgetsService', useClass: WidgetsService }
            ]
        };
    }
}
DashboardModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    DashboardComponent,
                    PanelComponent,
                    SpacerComponent,
                    PanelControlsComponent,
                    WidgetComponent,
                    WidgetControlsComponent,
                    LibraryComponent,
                    DropTargetComponent,
                    ResizeGuidelineComponent,
                    LayoutEditorComponent
                ],
                imports: [BrowserModule],
                exports: [
                    DashboardComponent,
                    PanelComponent,
                    SpacerComponent,
                    PanelControlsComponent,
                    WidgetComponent,
                    WidgetControlsComponent,
                    LibraryComponent,
                    DropTargetComponent,
                    ResizeGuidelineComponent,
                    LayoutEditorComponent
                ],
                providers: [BootstrapService],
                bootstrap: [DashboardComponent]
            },] },
];
function DashboardModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    DashboardModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    DashboardModule.ctorParameters;
}
